### PR TITLE
Add build operation progress events when builds in the tree and their projects are identified

### PR DESCRIPTION
**Pull request by adammurdoch**
Fixes #?

### Context

These events allow the structure of the build tree to be collected in a way that does not rely on particular build operations to produce that structure. This will allow those build operations to differ for builds where configuration caching is not enabled, when there is a cache miss and when there is a cache hit.

This PR is preparation work for https://github.com/gradle/gradle/issues/21985

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

**Comment by adammurdoch** [comment]
@bot-gradle test and merge

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt [hunk 1]
```diff
@@ -754,6 +755,7 @@ class ConfigurationCacheState(
         BuildOperationFiringSettingsPreparer(
```
Comment (guylabs) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:755]: Can't these synthetic build operations be removed as written in the spec ?

**Review by guylabs — APPROVED** [review]

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt (hunk 1 above)
Comment (adammurdoch) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/ConfigurationCacheState.kt:755]: Yes, that will be in the next set of changes. I just wanted to do the minimum to add the new events.

**Pull request merged**